dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/websockets/WsConnection.scala

```diff
@@ -1,64 +1,138 @@
 package com.wavesplatform.dex.it.api.websockets
 
+import java.util.UUID
 import java.util.concurrent.ConcurrentLinkedQueue
 
 import akka.Done
-import akka.actor.ActorSystem
+import akka.actor.{Actor, ActorRef, ActorSystem, Props, Status}
 import akka.http.scaladsl.Http
 import akka.http.scaladsl.model.HttpHeader
-import akka.http.scaladsl.model.ws.{Message, WebSocketRequest, WebSocketUpgradeResponse}
-import akka.stream.Materializer
-import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
+import akka.http.scaladsl.model.ws.{Message, TextMessage, WebSocketRequest, WebSocketUpgradeResponse}
+import akka.stream.scaladsl.{Flow, Sink, Source}
+import akka.stream.{CompletionStrategy, Materializer, OverflowStrategy}
+import cats.syntax.option._
 import com.wavesplatform.dex.api.http.`X-Api-Key`
+import com.wavesplatform.dex.api.websockets.WsMessage
+import com.wavesplatform.dex.api.websockets.actors.PingPongHandler.PingOrPong
 import com.wavesplatform.dex.domain.utils.ScorexLogging
+import com.wavesplatform.dex.it.api.websockets.WsConnection.SourceProxyActor
+import play.api.libs.json.Json
 
 import scala.collection.JavaConverters._
-import scala.concurrent.{Future, Promise}
-import scala.util.{Failure, Success}
+import scala.concurrent.Future
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success, Try}
 
-class WsConnection[Output](uri: String, parseOutput: Message => Output, trackOutput: Boolean, apiKey: Option[String] = None)(
-    implicit system: ActorSystem,
-    materializer: Materializer)
+class WsConnection[Output <: WsMessage: ClassTag](uri: String,
+                                                  parseOutput: Message => Output,
+                                                  trackOutput: Boolean,
+                                                  apiKey: Option[String] = None,
+                                                  keepAlive: Boolean = true)(implicit system: ActorSystem, materializer: Materializer)
     extends ScorexLogging {
 
-  log.info(s"Connecting to ws://$uri, apiKey = $apiKey")
+  log.info(s"""Connecting to Matcher WS API:
+            |         URI = ws://$uri
+            |     API Key = $apiKey
+            |  Keep alive = $keepAlive""".stripMargin)
+
+  private val sourceProxy = system.actorOf(SourceProxyActor.props(keepAlive))
+
+  private val pongHandler: Source[TextMessage.Strict, ActorRef] = {
+
+    val completionMatcher: PartialFunction[Any, CompletionStrategy] = { case akka.actor.Status.Success(_) => CompletionStrategy.draining }
+    val failureMatcher: PartialFunction[Any, Throwable]             = { case Status.Failure(cause)        => cause }
+
+    Source
+      .actorRef[PingOrPong](completionMatcher, failureMatcher, 10, OverflowStrategy.fail)
+      .map(_.toStrictTextMessage)
+      .mapMaterializedValue { source =>
+        sourceProxy.tell(SourceProxyActor.ProxyRequest, source)
```

Comment:
It _is_ proxy, since we cannot send pongs to `pongHandler` directly. We can do that only via `sourceProxy` actor

Reply:
[Proxy](https://en.wikipedia.org/wiki/Proxy_pattern#UML_class_and_sequence_diagram) is when all requests pass through it. But in the code:
```scala
      .mapMaterializedValue { source =>
        sourceProxy.tell(SourceProxyActor.ProxyRequest, source)
        source // we use an original actor.
        // This will be a proxy if we return sourceProxy instead
      }
```

Reply:
Now:
`pongHandler` is a `source` 
`sourceProxy` is a `pongHandler`